### DISCUSSION TITLE
Access QtSignals from external module / script

### DISCUSSION THREAD

**Discussion by AndrePointner** [Ulm-IQO/qudi-core#48] [Q&A] [answered]
Hello everyone,

I struggle to find out if and how it is possible to access QtSignals from an external python script to controll QUDI functionality from an external script. I'd like to incorporate some QUDI functionality like POIs and the Optimizer in Python Code that controls other non-QUDI hardware.

Is it possible to connect a "fresh" python instance to the pyqt Signals / Threads? I started using the jupyter kernel as a "workaround" by sending manual jupyter commands via a python script, but this feels... wrong.

Thank you for any help!

Best wishes

**Comment by liuyanulm** [comment]
This is might be dangerous… better call certain functions in the logic
module or hardware module..

Hundsraudigerrotzbua ***@***.***>于2023年4月25日 周二17:26写道：

**Reply by AndrePointner** [reply]
How would I go about this? I can't just import a function from a logic module into a python script and call it. How do I ensure a connection to the running QUDI Application?

**Comment by liuyanulm** [comment]
You can import the related module ie xxx, and call the function
xxx.function (), just like you use other python modules.

Hundsraudigerrotzbua ***@***.***>于2023年4月25日 周二18:48写道：

**Reply by AndrePointner** [reply]
Can you elaborate further? 

If I want to simple start and stop the time series reading with start_reading() / stop_reading() in the time_series_reader_logic.py, how would I got about importing this in a unrelated python script?

I can't simply import the logic class in a new script and launch the function, as there is no connection to the running QUDI thread, is there?

**Comment by Neverhorst** [comment]
Interfacing with a running qudi instance from a jupyter notebook can be done with the following steps:

1. Run qudi and start all required modules
2. Open a command line and activate the qudi Python environment (different for "Anaconda" and "Python for Windows")
3. Start a jupyter notebook from the same command line by typing e.g. `jupyter-notebook`
4. In the browser window that has opened, start a new notebook choosing the kernel named "qudi" or open an existing notebook and change the kernel to "qudi".
![image](https://user-images.githubusercontent.com/18266223/236809859-b0b82f45-c460-4144-b87a-744b97257c48.png)

5. you now have all the currently active qudi modules in the notebook namespace. The name of the modules is the one configured in the qudi config and shown on the "activate buttons" inside the qudi main window.
6. Call methods from logic modules. Avoid hardware and gui module access from notebooks... it works in principle but easily breaks depending on what you do.


Assuming you have loaded the `qudi-iqo-modules` default config file, starting and stopping the time series reader could look like this:
![image](https://user-images.githubusercontent.com/18266223/236810746-25b67826-c49a-4137-8c01-a136b499a50c.png)

**Reply by Neverhorst** [reply]
If you do not want to use jupyter notebooks but plain python scripts, it is much more difficult but doable.
You would then need to set up a local rpyc client connection to the qudi namespace server in your script and request the activated module proxy objects manually to call methods on them.

But I would advise against it. In this case it would be much easier to just go ahead and write a qudi logic module that contains the desired functionality.

**Comment by Neverhorst** [comment]
@Hundsraudigerrotzbua , did any of the answers above help you or do you need more info? Please mark an answer if it was helpful or post your solution to the problem.

**Reply by AndrePointner** [reply, marked as answer]
I've been successfully using the rpyc client connection from outside qudi. Took me a bit of tinkering, as the documentation wasn't too explicit, but managed to get it working like this:

```
class qudi_remote_control():

    def __init__(self, host='localhost', port=12345, conn_config={'allow_all_attrs': True}):
        self.host = host
        self.port = port
        self.conn_config = conn_config
        self.connection = rpyc.connect(self.host, self.port, config=self.conn_config)

    def start_timseries(self, *args, **kwargs):
        tm = self.connection.root.getModule('timeserieslogic')
        tm.start_reading()
```
Its neither the prettiest and might be prone to errors, but it works like a charm and does the job for now.

Thanks for the suggestion.
